File: Exercise_3.cpp
/* 
    Time Complexity                              :  O(3*N)
    Space Complexity                             :  O(1)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  


class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size();
        
        int i = n-2;
        while(i >= 0 and nums[i+1] <= nums[i]) {
            i--;
        }
        
        if(i >= 0) {
            // swap(nums[i],nums[i+1]);
            int j = n-1;
            while(j>0 and nums[j] <= nums[i]) j--;
            swap(nums, i, j);
        }
        
        reverse(nums,i+1);
    }
    
    void reverse(vector<int>& nums, int l) {
        int r = nums.size()-1;
        while(l < r) {
            swap(nums, l, r);
            l++;
            r--;
        }
    }
                
    void swap(vector<int>& nums, int &i, int &j) {
        int temp = nums[j];
        nums[j] = nums[i];
        nums[i] = temp;

    }
};

File: Exercise_3.py
# Problem3  Next permutation(https://leetcode.com/problems/next-permutation/)
# // Time Complexity :O(n)
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No

class Solution:
    def nextPermutation(self, nums):
        """
        Do not return anything, modify nums in-place instead.
        """
        if not nums:
            return 0
        n = len(nums)
        i = n - 2
        # 1. find breach
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1
        # 2. swap
        if i != -1:
            j = n - 1
            while nums[i] >= nums[j]:
                j -= 1
            self.swap(nums, i, j)
        # 3. reverse from i to end
        self.reverse(nums, i + 1, n - 1)

    def swap(self, nums, i, j):
        temp = nums[i]
        nums[i] = nums[j]
        nums[j] = temp

    def reverse(self, nums, l, r):
        while l < r:
            self.swap(nums, l, r)
            l += 1
            r -= 1


sol = Solution()
print(sol.nextPermutation([1, 2, 3]))
